Support tflite kernel unittest in cmake.

JerryShih opened the pull request:
We don't have the kernel test case with cmake build.
This pr try to port the bazel setting into cmake for the kernel tests.

Here are the output message for the tests:
```
/tensorflow/tensorflow/lite/build$ ctest

Test project /tensorflow/lite/build
        Start   1: internal-averagepool_quantized_test
  1/137 Test   #1: internal-averagepool_quantized_test ......................   Passed    0.18 sec
        Start   2: internal-batch_to_space_nd_test
  2/137 Test   #2: internal-batch_to_space_nd_test ..........................   Passed    0.00 sec
        Start   3: internal-conv_per_channel_quantized_16x8_test
  3/137 Test   #3: internal-conv_per_channel_quantized_16x8_test ............   Passed    1.24 sec
  ....
        Start 137: test_util_test
137/137 Test #137: test_util_test ...........................................   Passed    0.00 sec

100% tests passed, 0 tests failed out of 137
```

JerryShih committed:
Update tflite versioning UpdateOpVersion() interface.

For cross-compiling case, we can't generate the mutable schema easily.
It's complicated to run the target-specific flatc executable on the host
side. So, we can't get the mutable interfaces used in UpdateOpVersion().

This commit checks a definition named "TFLITE_VERSION_USE_MUTABLE_SCHEMA"
for using the immutable and mutable schema header.
If we use the immutable interface, use pack/unpack mechanism to update the
buffer content in the builder

JerryShih committed:
Update the tflite model "buffers" field checking rule.

If we don't use "--force-empty-vectors" flag[1] for flatc, the buffers
might be a null ptr if we serialize a model with zero buffers size(e.g.
all ops in model doesn't use const weights in model).
This commit relaxs the "buffers" null ptr checking for this situation,
and also updates the "subgraphs" checking for null ptr dereference.

[1]
https://github.com/google/flatbuffers/blob/fc4fffea41f5b682198edfc72eca536d33fd848c/src/flatc.cpp#L167

JerryShih committed:
Support tflite kernel test in cmake.

1. Add gtest, nsync and re2 library.
2. Add most of ".*_test.cc" test cases in tflite kernels folder.
3. Add TFLITE_KERNEL_TEST option for kernel testing on/off.

How to run tflite kernel test with ctest:
1. Turn on TFLITE_KERNEL_TEST optoin in "tensorflow/lite/CMakeLists.txt"
2. Build tflite with cmake. [1]
3. Run "ctest" in the cmake build folder.

[1]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/g3doc/guide/build_cmake.md

JerryShih committed:
Fix gcc 5.4.0 compiling error.

The isnan() is not in global namespace.

JerryShih committed:
Fix gcc 5.4.0 compiling error.

Error message:
  error: narrowing conversion of ‘2147483647’ from ‘int’ to ‘float’ inside { } [-Wnarrowing]

JerryShih committed:
Skip the flatbuffers test cases in cmake.

JerryShih commented:
@terryheo @mihaimaruseac 
I update the UpdateOpVersion() interface in this pr. You guys might be worried about it.
I show the the detail reason and the implementation in the comment.
Please check:
https://github.com/tensorflow/tensorflow/pull/47754/commits/31f79ad804cccc071b030a3950106558edd8b30e
https://github.com/tensorflow/tensorflow/commit/9dc7dc24685087e6f5d0806966cb00176449360e

JerryShih commented on tensorflow/lite/tools/cmake/modules/Findgoogletest.cmake:
```diff
@@ -0,0 +1,29 @@
+
+include(FetchContent)
+
+FetchContent_Declare(
```
@terryheo 
Should we always use OverridableFetchContent_Declare()?

terryheo replied:
I prefer to use it for the consistency.

JerryShih replied:
Ok, I will turn to use OverridableFetchContent_Declare() in next commit.

JerryShih replied:
@terryheo 
This is the commit using OverridableFetchContent_Declare().
9afaf4824411c1ad3ea9730e318b82e49b6cd307

JerryShih committed:
Use OverridableFetchContent_Declare() instead of FetchContent_Declare() for tflite library.

Use OverridableFetchContent_Declare() for googletest, nsync and re2 libraries.

terryheo reviewed (commented):
thanks for your contribution.

JerryShih commented on tensorflow/lite/tools/versioning/op_version.cc:
```diff
@@ -858,30 +857,54 @@ OpSignature GetOpSignature(const OperatorCode* op_code, const Operator* op,
 
-void UpdateOpVersion(uint8_t* model_buffer_pointer) {
-  auto model = GetMutableModel(model_buffer_pointer);
+void UpdateOpVersion(flatbuffers::FlatBufferBuilder& builder) {
```
@terryheo 
Please check:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/test_util.cc#L193
The kernel tests need this function to update the op's version.
All kernel test's default op version is 0. If we don't update the op version, the interpreter can't not find the proper op.

JerryShih replied:
> I see. We used to have a rule to invoke flatc but removed it since it didn't work well on cross compilation.
> Let me find a way to bundle mutable/schema_generated.h file.

@terryheo 
Could we also commit the mutable/schema_generated.h just as the immutable one in repo? When I check the log of immutable schema header, I saw it's be updated if the schema changes. How about just upload both the immutable and mutable headers?

JerryShih commented on tensorflow/lite/CMakeLists.txt:
```diff
@@ -33,6 +33,14 @@ if(NOT CMAKE_BUILD_TYPE)
 endif()
 
+# ccache
+find_program(CCACHE_FOUND ccache)
```
I will do that later.

JerryShih replied:
@terryheo 
Here is the pr for ccache.
https://github.com/tensorflow/tensorflow/pull/47850

JerryShih commented on tensorflow/lite/interpreter_builder.cc:
```diff
@@ -564,11 +564,13 @@ TfLiteStatus InterpreterBuilder::ParseTensors(
+      if (tensor->buffer() == 0) {
+        return kTfLiteOk;
+      }
+      if (!buffers || tensor->buffer() >= buffers->size()) {
```
I show the detail in the comment.
https://github.com/tensorflow/tensorflow/pull/47754/commits/582bf0d3ac33fc10156f737c0d42f3adee54409a

If we build with cmake, we can't generate the mutable schema header for the original UpdateOpVersion() call. So, we can't call the UpdateOpVersion() in cmake build.

If I change the UpdateOpVersion() implementation using the pack/unpack with immutable schema header, I need the additional --force-empty-vectors flag for the immutable header. Otherwise, some vector field will become null-ptr if the vector size is 0 for serialization.
I don't regenerate the immutable schema header here, so I need to relax the buffers checking in interpreter_builder.cc. Some kernel tests doesn't have buffer. The buffer field will become null-ptr if the model comes from my new UpdateOpVersion() pack/unpack implementation.

JerryShih commented on tensorflow/lite/kernels/CMakeLists.txt:
```diff
@@ -0,0 +1,235 @@
+endmacro()
+
+# The tflite kernel test
+set(TEST_LIST
```
@terryheo
I need to do some special handling for the following test.
Some tests need to be skipped and some need to use different link options.
Is it good to use populate_source_vars() with complicated filter rule?
 
```
# We don't always have the delegate environment(e.g nnapi)
# So, skip this test.
#test_delegate_providers_test.cc

# The test which has main() function.
set(TEST_WITH_MAIN_LIST
  cpu_backend_gemm_test.cc
  cpu_backend_threadpool_test.cc
  eigen_support_test.cc
  kernel_util_test.cc
  optional_tensor_test.cc
  subgraph_test_util_test.cc
  test_util_test.cc
)
```

JerryShih committed:
Use host-side flatc compiler to generate the mutable schema interface for tflite kernel tests with cmake.

For cross compilation, we can't run the flatc built from cross-toolchain.
We could use cmake ExternalProject_Add() to build the flatc with host-side
compiler. Then, we could generate the mutable interface on-the-fly for
cross build.

JerryShih committed:
Revert "Update the tflite model "buffers" field checking rule."

This reverts commit 582bf0d3ac33fc10156f737c0d42f3adee54409a.

JerryShih committed:
Revert "Update tflite versioning UpdateOpVersion() interface."

This reverts commit 31f79ad804cccc071b030a3950106558edd8b30e.

JerryShih commented:
@terryheo 
Could you please check this commit for the mutable schema interface?
https://github.com/tensorflow/tensorflow/pull/47754/commits/bb088b080a72e790c18e8164ce9fb828b2cec5e4
It builds the host-side flatc using cmake ExternalProject_Add(). Then, we could use the flatc to generate the mutable schema interface.

terryheo commented on tensorflow/lite/kernels/CMakeLists.txt:
```diff
@@ -17,7 +17,26 @@ find_package(googletest REQUIRED)
+set(SCHEMA_GENERATED_ROOT ${CMAKE_CURRENT_BINARY_DIR}/schema)
+set(SCHEMA_GENERATED_PATH ${SCHEMA_GENERATED_ROOT}/tensorflow/lite/schema/mutable)
+set(SCHEMA_GENERATED_FILE ${SCHEMA_GENERATED_PATH}/schema_generated.h)
+add_custom_command(
```
Can't we use build_flatbuffers() in https://github.com/google/flatbuffers/blob/master/CMake/BuildFlatBuffers.cmake ?

JerryShih replied:
> Can't we use build_flatbuffers() in https://github.com/google/flatbuffers/blob/master/CMake/BuildFlatBuffers.cmake ?

I need to check if this is available using ExternalProject_Add() flow.

JerryShih replied:
@terryheo 
Here is the commit to use build_flatbuffers().
https://github.com/tensorflow/tensorflow/pull/47754/commits/809468d4407beaf80064b90f8c6d945de6394070

terryheo commented on tensorflow/lite/tools/cmake/modules/flatbuffers.cmake:
```diff
@@ -42,3 +42,18 @@ add_subdirectory(
+# The host-side flatc binary
+include(ExternalProject)
+
+ExternalProject_Add(flatbuffers-flatc
```
Oh I'm looking at this too. Did you verify it worked well with cross compilation?

JerryShih replied:
@terryheo 
> Oh I'm looking at this too. Did you verify it worked well with cross compilation?

Yes, I use the cross-tool for tflite and the host-tool for flatc.
I could see the mutable interface was generated successfully.
```
/tensorflow/tensorflow/lite/tools/cmake/build$ ls kernels/schema/tensorflow/lite/schema/mutable/schema_generated.h
kernels/schema/tensorflow/lite/schema/mutable/schema_generated.h
```

JerryShih committed:
Use the util function in flatbuffer library to generate the schema header.

terryheo commented on tensorflow/lite/CMakeLists.txt:
```diff
@@ -70,6 +70,9 @@ option(TFLITE_ENABLE_GPU "Enable GPU" OFF)
 option(TFLITE_ENABLE_FLEX "Enable SELECT_TF_OPS" OFF) # TODO: Add support
 option(TFLITE_ENABLE_XNNPACK "Enable XNNPACK backend" ON)
+
+option(TFLITE_KERNEL_TEST "Enable tflite kernel unit test" ON)
```
Shall we use "OFF" for default?

JerryShih replied:
@terryheo 
Does tensorflow have cmake-base ci test?
If yes, is it better to run the kernel tests for cmake build?

terryheo replied:
Yes we have. But it's different issue. We can always control this by providing the CMake argument.
I think most users don't want to build it by default.

JerryShih replied:
Oh.. Let's control it by the cmake argument.

terryheo commented on tensorflow/lite/kernels/CMakeLists.txt:
```diff
@@ -0,0 +1,262 @@
+  sub_test.cc
+  svdf_test.cc
+  # We don't always have the delegate environment(e.g nnapi)
+  # So, skip this test.
```
So, skip "test_delegate_providers_test.cc".

terryheo commented on tensorflow/lite/kernels/quantize_test.cc:
```diff
@@ -482,7 +482,7 @@ TEST(QuantizeOpTest, Int16Int32SameScale) {
                      {1, 1, 2, 5},
                      std::numeric_limits<int32_t>::min(),
-                     std::numeric_limits<int32_t>::max()});
+                     (float)std::numeric_limits<int32_t>::max()});
```
Let's use "static_cast<float>()"

terryheo commented on tensorflow/lite/kernels/quantize_test.cc:
```diff
@@ -501,7 +501,7 @@ TEST(QuantizeOpTest, Int16Int32LargerScale) {
                      {1, 1, 2, 5},
                      std::numeric_limits<int32_t>::min(),
-                     std::numeric_limits<int32_t>::max()});
+                     (float)std::numeric_limits<int32_t>::max()});
```
ditto

terryheo commented on tensorflow/lite/CMakeLists.txt:
```diff
@@ -421,6 +424,13 @@ target_compile_options(tensorflow-lite
 )
 add_library(tensorflow::tensorflowlite ALIAS tensorflow-lite)
 
+enable_testing()
```
shouldn't be inside of "if(TFLITE_KERNEL_TEST)" ?

terryheo commented on tensorflow/lite/kernels/CMakeLists.txt:
```diff
@@ -0,0 +1,262 @@
+set(FLATBUFFERS_FLATC_EXECUTABLE ${CMAKE_BINARY_DIR}/flatbuffers-flatc/bin/flatc)
+set(FLATBUFFERS_FLATC_SCHEMA_EXTRA_ARGS
+  -c
+  --no-union-value-namespacing
```
is this needed?

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/schema/BUILD#L108

JerryShih replied:
@terryheo 
Oh... I misunderstood the flatbuffer_cc_library() call.
https://github.com/tensorflow/tensorflow/blob/38fca54c85764aadb40f7999ce14aa04b29d8653/third_party/flatbuffers/build_defs.bzl#L15

The default option will be replaced by user.
I will remove the "no-union-value-namespacing" option in next commit.

terryheo reviewed (commented):
It's nice!

JerryShih committed:
Update cmake setting for review comment.

JerryShih committed:
Update kernel cmake setting for review comment.

JerryShih committed:
Use c++ style casting.

JerryShih commented:
@terryheo 
All review comments are addressed.
Do I need to squash the fixing commits?

terryheo commented:
> Do I need to squash the fixing commits?
No, it's fine. This PR will be merged as a single commit with history.

terryheo reviewed (approved):
Thanks for your contribution! It's looks fine to me.
But we need another component owner's review.
I'll share it if there is anything to address.

multiverse-tf commented:
Really appreciate the contribution!
